lib/crypto.js

```diff
@@ -95,6 +96,28 @@ Hmac.prototype.digest = Hash.prototype.digest;
 Hmac.prototype._flush = Hash.prototype._flush;
 Hmac.prototype._transform = Hash.prototype._transform;
 
+// This implements Brad Hill's Double HMAC pattern from
+// https://www.nccgroup.trust/us/about-us/
+//   newsroom-and-events/blog/2011/february/double-hmac-verification/.
+// In short, it's near-impossible to write a reliable constant-time compare in a
+// high level language like JS, because of the many layers that can optimize
+// away attempts at being constant time.
+//
+// Double HMAC avoids that problem by blinding the timing channel instead. After
+// running the inputs through a second round of HMAC, we are free to
+// short-circuit comparison, because the time it takes to reach the
+// short-circuit has no relation to the similarity between the guessed digest
+// and the correct one.
+//
+// Note: hmac object can not be used after validate() method has been called.
+Hmac.prototype.validate = function(inputBuffer) {
+  if (!(inputBuffer instanceof Buffer)) {
+    throw new TypeError('Argument should be a Buffer');
+  }
+  var ah = new Hmac('sha256', this.key).update(this.digest()).digest();
+  var bh = new Hmac('sha256', this.key).update(inputBuffer).digest();
+  return ah.equals(bh);
```

Comment:
> What benefits does this provide over just using timingSafeEqual?

Not much. It follows Brad Hill's original construction a little more closely, and it avoids generating a random number. But the simplicity seems worthwhile: would you like me to change it to that way?